### DISCUSSION TITLE
TMS Integration with GeoServer

### DISCUSSION THREAD

**Discussion by fcdalgic** [pelicanmapping/osgearth#2653] [Support Q&A] [answered]
Hello, 
I am newbie to osgEarth and GeoServer and trying to integrate GeoServer's TMS service with osgEarth's latest version. I have already successfully managed to integrate with WMS service to retrieve raster data. However, when I modified the example readymap.earth file to set TMSElevationLayer's URL with "http://localhost:8080/geoserver/RsMap/gwc/service/tms/1.0.0/RsMap%3ARsTiles@global-geodetic@png", it gives me some errors:

#### Resources

1. The content of the earth file:
```
<map name="ReadyMap.org">

 
	 <TMSElevation name="ReadyMap 90m Elevation">
		 <url>http://localhost:8080/geoserver/RsMap/gwc/service/tms/1.0.0/RsMap%3ARsTiles@global-geodetic@png</url>
		 <vdatum>egm96</vdatum>
	 </TMSElevation>

    <xi:include href="viewpoints.xml"/>
</map>
```

2. On the HTTP Debug console is attached. [console_output.txt](https://github.com/user-attachments/files/17679001/console_output.txt)
3. The actual response from TMS is attached. [TMSResponse.xml.txt](https://github.com/user-attachments/files/17679003/TMSResponse.xml.txt)

#### Questions:

1. Why TMSElevation tries to retrieve data from "http://localhost:8080/geoserver/RsMap/gwc/service/tms/1.0.0/0/0/0.png" instead of  "http://localhost:8080/geoserver/RsMap/gwc/service/tms/1.0.0/RsMap%3ARsTiles@global-geodetic@png/0/0/0.png"?
2. How to integrate with GeoServer's TMS service?
3. Is TIF the only supported format for TMSElevationLayer?
4. If I needed manipulate the service parser, which class or method should I manipulate to cope it?

Thanks in advance,

**Comment by gwaldron** [comment, marked as answer]
We reproduced your problem, converted this to an issue, and pushed a patch #2655

**Reply by canacar7** [reply]
Hi,

In the Geoserver application, I group DTED data by creating layers. I want to retrieve these group layers from the TMS server. However, Geoserver can convert DTED data into PNG or JPEG formats.

When I use TMSElevation in osgEarth to retrieve the data, I cannot use it as elevation data. Can I use data in PNG format as elevation data? Could you please help me with this?

Thank you.

**Reply by gwaldron** [reply]
osgEarth supports RGB elevation data using the `XYZElevationLayer` only at this time. We could add support to TMS as well if necessary.
We support two RGB elevation data encodings: MapBox Terrain-RGB ("mapbox") and MapZen Terrarium ("terrarium").
Please refer to the `mapzen.earth` sample in the `tests` folder for an example.

**Reply by canacar7** [reply]
<XYZElevation name="elevation">
        <url>http://localhost:8080/geoserver/EarthWS/gwc/service/tms/1.0.0/turkey_st_roads@global_geodetic@png</url>
        <elevation_encoding>terrarium</elevation_encoding>
        <stitch_edges>true</stitch_edges>
      </XYZElevation> 

But you're getting the error: "XmlInputIterator::readProperty(): Unmatched property #SRS, expecting #Generator".

What would be the best way to resolve this issue? Could you help me figure out the problem and how to fix it?

**Reply by gwaldron** [reply]
You are attempting to use the `XYZElevation` layer to read a TMS data source. You need to use a `TMSElevation` layer instead. (Make sure you build the latest master branch of osgEarth to get the `elevation_encoding` support for TMS.)

You can use XYZ, but you would need to access the tiles directly using the {z}/{x}/{y} idiom; see the openstreetmap_flat.earth for an example.

**Reply by canacar7** [reply]
Hi, I added my data in DTED format as an imageMosaic in the GeoServer application. Then, I configured the gridset settings and generated TMS layers in the tileLayer section. The output format of the data was saved as TIFF. When I add the data to the map using TMSImage, it looks quite good. However, when I try to add it with TMSElevation, the result is not successful. I am sharing the .earth file and screenshots I prepared for both the image and elevation forms.
![mapElevation2](https://github.com/user-attachments/assets/eaa564a4-75f0-4595-991e-8e4fbb91e8c5)
![mapImage2](https://github.com/user-attachments/assets/b07202c6-d01e-46cf-9762-afe61973f2da)


![earthElevation3](https://github.com/user-attachments/assets/c94dd3a3-b90a-4f9e-ac34-08a1ad584056)
![eartImage3](https://github.com/user-attachments/assets/451be7b1-f188-43f8-a57e-b904821716c2)

**Comment by fcdalgic** [comment]
Thank you, in the meantime, I might be able help with that if you could guide me, thanks
